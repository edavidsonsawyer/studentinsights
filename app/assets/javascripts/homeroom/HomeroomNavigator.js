import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import Select from 'react-select';
import * as Routes from '../helpers/Routes';
import {ORDERED_GRADES} from '../helpers/gradeText';
import Button from '../components/Button';


// For typing a homeroom name and jumping to it, or dropping down a list and clicking.
// Requires clicking "Go" to navigate.
export default class HomeroomNavigator extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      homeroomOption: null
    };
    this.onHomeroomItemChanged = this.onHomeroomItemChanged.bind(this);
    this.onNavigate = this.onNavigate.bind(this);
  }

  onHomeroomItemChanged(homeroomOption) {
    this.setState({homeroomOption});
  }

  onNavigate() {
    const {homeroomOption} = this.state;
    const id = homeroomOption.value;
    window.location = Routes.homeroom(id);
  }

  render() {
    const {homerooms, style} = this.props;
    const {homeroomOption} = this.state;

    const sortedHomerooms = _.sortBy(homerooms, homeroom => {
      const gradesSortKey = (homeroom.grades.length === 1)
        ? ORDERED_GRADES[homeroom.grades[0]]
        : ORDERED_GRADES.length + 1;
      return [gradesSortKey, homeroom.name];
    });
    const options = sortedHomerooms.map(homeroom => {
      return {
        value: homeroom.id,
        label: `${homeroom.name} (${homeroomGradesText(homeroom)})`
      };
    });

    return (
      <div className="HomeroomNavigator" style={{...styles.root, ...style}}>
        <Select
          placeholder="Find homeroom..."
          value={homeroomOption}
          style={styles.select}
          onChange={this.onHomeroomItemChanged}
          options={options}
        />
        <Button
          isDisabled={homeroomOption === null}
          onClick={this.onNavigate}
        >Go</Button>
      </div>
    );
  }
}
HomeroomNavigator.propTypes = {
  homerooms: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    grades: PropTypes.arrayOf(PropTypes.string).isRequired
  })).isRequired,
  style: PropTypes.object
};


const styles = {
  root: {
    display: 'flex',
    flexDirection: 'row'
  },
  select: {
    width: '15em',
    marginRight: 10
  }
};


function homeroomGradesText(homeroom) {
  return (homeroom.grades.length === 1) ? homeroom.grades[0] : 'mixed';
}