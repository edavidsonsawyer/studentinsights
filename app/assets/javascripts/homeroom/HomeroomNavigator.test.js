import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import HomeroomNavigator from './HomeroomNavigator';
import homeroomJson from './homeroomJson.fixture';

function testProps(props = {}) {
  return {
    homerooms: homeroomJson.homerooms,
    ...props
  };
}

function testRender(props = {}, context = {}) {
  const el = document.createElement('div');
  ReactDOM.render(<HomeroomNavigator {...props} />, el);
  return {el};
}

it('renders without crashing', () => {
  const props = testProps();
  const {el} = testRender(props);
  expect($(el).html()).toContain('Find homeroom...');
  expect($(el).html()).toContain('button');
});

it('snapshots view', () => {
  const props = testProps();
  const tree = renderer
    .create(<HomeroomNavigator {...props} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});