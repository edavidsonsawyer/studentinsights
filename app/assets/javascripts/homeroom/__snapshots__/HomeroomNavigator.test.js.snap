// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots view 1`] = `
<div
  className="HomeroomNavigator"
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
    }
  }
>
  <div
    className="Select is-clearable is-searchable Select--single"
  >
    <div
      className="Select-control"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "marginRight": 10,
          "width": "15em",
        }
      }
    >
      <div
        className="Select-multi-value-wrapper"
        id="react-select-3--value"
      >
        <div
          className="Select-placeholder"
        >
          Find homeroom...
        </div>
        <div
          className="Select-input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-activedescendant="react-select-3--value"
            aria-expanded="false"
            aria-haspopup="false"
            aria-owns=""
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "5px",
              }
            }
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          >
            
          </div>
        </div>
      </div>
      <span
        className="Select-arrow-zone"
        onMouseDown={[Function]}
      >
        <span
          className="Select-arrow"
          onMouseDown={[Function]}
        />
      </span>
    </div>
  </div>
  <div
    className="Hover"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={Object {}}
  >
    <button
      style={
        Object {
          "background": "#aaa",
          "border": "none",
          "borderRadius": 28,
          "color": "#eee",
          "cursor": "default",
          "fontSize": 14,
          "padding": "8px 25px",
          "textDecoration": "none",
        }
      }
    >
      Go
    </button>
  </div>
</div>
`;
